staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/kms/kms.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package kms transforms values for storage at rest using a KMS provider
+package kms
+
+import (
+	"crypto/aes"
+	"crypto/rand"
+	"encoding/binary"
+	"fmt"
+	"sync"
+
+	"k8s.io/apiserver/pkg/storage/value"
+	aestransformer "k8s.io/apiserver/pkg/storage/value/encrypt/aes"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+type kmsTransformer struct {
+	kmsService   value.KMSService
+	transformers *lru.Cache
+
+	// cacheSize is the maximum number of DEKs that are cached.
+	cacheSize int
+
+	lock sync.RWMutex
+}
+
+// NewKMSTransformer returns a transformer which implements a KEK-DEK based envelope encryption scheme.
+// It uses kmsService to communicate with the KEK store, and storage to communicate with the DEK store.
+func NewKMSTransformer(kmsService value.KMSService, cacheSize int) (value.Transformer, error) {
+	cache, err := lru.New(cacheSize)
+	if err != nil {
+		return nil, err
+	}
+	return &kmsTransformer{
+		kmsService:   kmsService,
+		transformers: cache,
+		cacheSize:    cacheSize,
+	}, nil
+}
+
+// TransformFromStorage implements value.Transformer
```

Comment:
Could you change this comment to something that explains what is happening here. If you want this to be an implementation of an interface, add a unit test that assigns the result of this function to a variable of the interface type. Then the compiler will enforce the type for you.